contracts/src/mixins/QuotaMixin.sol

```diff
@@ -0,0 +1,200 @@
+        IERC20(FEE_TOKEN).safeTransferFrom(msg.sender, address(this), amount);
+        emit Deposit(signer, amount);
+    }
+
+    function computeWithdrawHash(
+        uint256 amount,
+        address beneficiary,
+        uint256 nonce
+    ) internal view returns (bytes32 withdrawHash) {
+        bytes32 domainSeparator = keccak256(
+            abi.encode(
+                keccak256(
+                    "EIP712Domain(uint256 chainId,address verifyingContract)"
+                ),
+                block.chainid,
+                address(this)
+            )
+        );
+        bytes32 structHash = keccak256(
+            abi.encode(
+                keccak256(
+                    "WithdrawRequest(uint256 amount,address beneficiary,uint256 nonce)"
+                ),
+                amount,
+                beneficiary,
+                nonce
+            )
+        );
+        withdrawHash = keccak256(
+            abi.encodePacked("\x19\x01", domainSeparator, structHash)
+        );
+    }
+
+    function widthdrawTokensForSigner(
+        bytes calldata signature,
+        uint128 amount,
+        address beneficiary,
+        uint256 nonce
+    ) public payable {
+        bytes32 withdrawHash = computeWithdrawHash(amount, beneficiary, nonce);
+        (address signer, bytes32 r, bytes32 vs) = CoreLib.recoverSigner(
+            withdrawHash,
+            signature
+        );
+        // Silence unused local variable warning
+        (r, vs);
+        // Check that withdrawal was not executed yet
+        require(withdrawsForSigner[signer][withdrawHash] == 0, "Withdrawal was already performed");
+        withdrawsForSigner[signer][withdrawHash] = block.timestamp;
+
+        QuotaStats storage stats = quotaStatsForSigner[signer];
+        require(stats.tokenBalance >= amount, "Insufficient Tokens");
+        // We use the quota reset timeframe as a unlock timeframe
+        // -> currently the signer is not allowed to sign any Safe transaction during this timeframe
+        // TODO: have dedicated unlock logic (also to avoid some fee exploit flows)
+        require(
+            stats.nextQuotaReset < block.timestamp,
+            "Tokens have been used during this timeframe"
+        );
+
+        stats.tokenBalance -= amount;
+
+        IERC20(FEE_TOKEN).safeTransfer(beneficiary, amount);
+        emit Withdraw(signer, amount);
+    }
+
+    function availableFreeQuotaForSigner(
+        address signer
+    )
+        public
+        view
+        returns (
+            uint64 availableFreeQuota,
+            uint64 usedSignerQuota,
+            uint64 nextSignerQuotaReset
+        )
+    {
+        QuotaStats memory stats = quotaStatsForSigner[signer];
+        nextSignerQuotaReset = stats.nextQuotaReset;
+        if (nextSignerQuotaReset > block.timestamp) {
+            usedSignerQuota = stats.usedQuota;
+        } else {
+            uint64 blocktime = uint64(block.timestamp);
```

Comment:
Wasn't clear to me at a first read, but `usedSignerQuota` is implicitly 0 in this case. Might be worth a comment (had to double back to check).